File: web/ViewModels/FeedViewModel.cs
using web.Models;
using web.Services;

namespace web.ViewModels
{
    public enum FeedState { Idle, Loading, LoadingMore, Success, Error }
    public class FeedViewModel
    {
        private readonly ApiService _apiService;

        public FeedState State { get; private set; } = FeedState.Idle;
        public List<FeedItemModel> FeedItems { get; private set; } = new();
        public string ErrorMessage { get; private set; } = string.Empty;
        public bool HasMore { get; private set; } = true;
        public bool IsRefreshing { get; private set; }
        public bool IsEmpty => FeedItems.Count == 0 & State == FeedState.Success;

        private int _currentPage = 1;
        public event Action? OnChange;
        
        public FeedViewModel(ApiService apiService)
        {
            _apiService = apiService;
        }

        public async Task FetchFeedAsync(bool refresh = false)
        {
            if (refresh)
            {
                _currentPage = 1;
                HasMore = true;
                IsRefreshing = true;
                FeedItems = new List<FeedItemModel>();
                NotifyStateChanged();
            }
            else
            {
                if (State == FeedState.Loading || State == FeedState.LoadingMore) return;

                SetState(FeedItems.Count == 0 ? FeedState.Loading : FeedState.LoadingMore);
            }

            try
            {
                var response = await _apiService.GetFeedAsync(_currentPage);
                var newItems = response.Items.Select(item => FeedItemModel.FromJson(item)).ToList();

                if (refresh)
                {
                    FeedItems = newItems;
                }
                else
                {
                    FeedItems = new List<FeedItemModel>(FeedItems);
                    FeedItems.AddRange(newItems);
                }
            }
            catch (Exception ex)
            {
                IsRefreshing = false;
                ErrorMessage = ex.Message;
                SetState(FeedState.Error);
            }
        }
        public async Task LoadMoreAsync() {
            if (!HasMore || State == FeedState.LoadingMore) return;
            await FetchFeedAsync();
        }
        public async Task RefreshAsync() {
            await FetchFeedAsync(refresh: true);
        }
        private void SetState(FeedState state)
        {
            State = state;
            NotifyStateChanged();
        }
        private void NotifyStateChanged() => OnChange?.Invoke();
    }
}

File: web/js/viewmodels/FeedViewModel.js
import { getFeed } from '../services/apiService';
import { FeedItem } from '../models/FeedItem';

export class FeedViewModel {
    #items = [];
    #state = 'idle';
    #currentPage = 1;
    #hasMore = true;
    #isRefreshing = false;
    #errorMessage = '';
    #listeners = new Set();

    get items() { return this.#items; }
    get state() { return this.#state; }
    get hasMore() { return this.#hasMore; }
    get isRefreshing() { return this.#isRefreshing; }
    get errorMessage() { return this.#errorMessage; }
    get isEmpty() { return this.#items.length === 0 && this.#state === 'success';}

    subscribe(fn) {
        this.#listeners.add(fn);
        return () => this.#listeners.delete(fn);
    }
    #notify() {
        this.#listeners.forEach((fn) => fn(this));
    }
    #setState(state) {
        this.#state = state;
        this.#notify();
    }

    async fetchFeed(refresh = false) {
        if(refresh) {
            this.#currentPage = 1;
            this.#hasMore = true;
            this.#isRefreshing = true;
            this.#items = [];
            this.#notify();
        } else {
            if(this.#state === 'loading' || this.#state === 'loadingMore') return;
            this.#setState(this.#items.length === 0 ? 'loading' : 'loadingMore');
        }

        try {
            const response = await getFeed(this.#currentPage);
            const newItems = FeedItem.fromList(response.items);

            this.#items = refresh ? newItems : [...this.#items, ...newItems];
            this.#hasMore = response.hasMore;
            this.#currentPage++;
            this.#isRefreshing = false;
            this.#setState('success');
        } catch(err) {
            this.#isRefreshing = false;
            this.#errorMessage = err.message;
            this.#setState('error');
        }
    }

    async loadMore() {
        if(!this.#hasMore || this.#state === 'loadingMore') return;
        await this.fetchFeed();
    }

    async refresh() {
        await this.fetchFeed(true);
    }
}